File: copilot-bot/src/commands/systemmessage.ts
import type { CommandHandler } from "@woife5/shared";
import {
    type ChatInputCommandInteraction,
    MessageFlags,
    PermissionFlagsBits,
    SeparatorBuilder,
    SeparatorSpacingSize,
    SlashCommandBuilder,
    TextDisplayBuilder,
} from "discord.js";
import { getSystemMessage, setSystemMessage } from "llm-connector/chat-history";

export const systemMessage: CommandHandler = {
    data: new SlashCommandBuilder()
        .setName("systemmessage")
        .setDescription("Edit the system message.")
        .setDefaultMemberPermissions(PermissionFlagsBits.Administrator)
        .addStringOption((option) => option.setName("message").setDescription("The new system message.")),
    executeInteraction: async (interaction: ChatInputCommandInteraction): Promise<void> => {
        const newMessage = interaction.options.getString("message");

        const separator = new SeparatorBuilder().setDivider(true).setSpacing(SeparatorSpacingSize.Large);

        if (!newMessage) {
            const header = new TextDisplayBuilder().setContent("Current system message:");
            const text = new TextDisplayBuilder().setContent(getSystemMessage());
            interaction.reply({
                components: [header, separator, text],
                flags: MessageFlags.IsComponentsV2,
            });
            return;
        }

        setSystemMessage(newMessage);
        const header = new TextDisplayBuilder().setContent("System message updated to:");
        const newText = new TextDisplayBuilder().setContent(newMessage);
        interaction.reply({ components: [header, separator, newText], flags: MessageFlags.IsComponentsV2 });
    },
};

File: copilot-bot/commands/systemmessage.go
package commands

import (
	"github.com/bwmarrin/discordgo"
)

func (b *Bot) handleSystemMessage(s *discordgo.Session, i *discordgo.InteractionCreate) {
	options := i.ApplicationCommandData().Options

	divider := true
	spacingLarge := discordgo.SeparatorSpacingSizeLarge
	separator := discordgo.Separator{
		Divider: &divider,
		Spacing: &spacingLarge,
	}

	// If no new message provided, show the current system message
	var newMessage string
	for _, opt := range options {
		if opt.Name == "message" {
			newMessage = opt.StringValue()
		}
	}

	if newMessage == "" {
		header := discordgo.TextDisplay{Content: "Current system message:"}
		text := discordgo.TextDisplay{Content: b.History.GetSystemMessage()}

		s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
			Type: discordgo.InteractionResponseChannelMessageWithSource,
			Data: &discordgo.InteractionResponseData{
				Components: []discordgo.MessageComponent{&header, &separator, &text},
				Flags:      discordgo.MessageFlagsIsComponentsV2,
			},
		})
		return
	}

	b.History.SetSystemMessage(newMessage)
	header := discordgo.TextDisplay{Content: "System message updated to:"}
	text := discordgo.TextDisplay{Content: newMessage}

	s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
		Type: discordgo.InteractionResponseChannelMessageWithSource,
		Data: &discordgo.InteractionResponseData{
			Components: []discordgo.MessageComponent{&header, &separator, &text},
			Flags:      discordgo.MessageFlagsIsComponentsV2,
		},
	})
}
